### PR TITLE
Forward proxy requests shouldn't resolve hostnames

### LINKED ISSUE #4244: skip the DNS query  if the proxy config  is config for the endpoint.
Config the client factory to use a proxy, if the URL could not resolve in localhost, an UnknownHostException would be thrown.

If we have config the client factory proxy, we should skip resolving the address for the proxy request.

### PR DESCRIPTION
**Pull request by jrhee17**
Motivation:

Say we connect to a destination which we will call `destinationAddress` via a proxy server with the address `proxyAddress`.
When a normal proxy client tries to make such a connection, it only needs to know the IP address of the `proxyAddress`.

However currently when an armeria proxy client tries to connect to `destinationAddress`, armeria first tries to resolve the IP address of `destinationAddress`, and then make a connection to `proxyAddress`. The resolution result is unnecessary and this step can be skipped for forward proxy requests.

It can be seen that DNS resolution is done irregardless of whether the request is a proxy request.
https://github.com/line/armeria/blob/790a2db01a8514b591bde90932ab18f63693829e/core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java#L95

Eventually, whether the request is a proxy is determined later when acquiring a connection.
https://github.com/line/armeria/blob/790a2db01a8514b591bde90932ab18f63693829e/core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java#L160

I propose that:
1) For forward proxy requests, dns resolution isn't done on host names (although users are still welcome to use IP addresses for `destinationAddress`)
2) `PoolKey` doesn't rely on `destinationAddress` IP for forward proxy requests. Now the connection will be keyed by `(destinationHost, destinationPort, proxyConfig)`. It might be worth discussing if it is really necessary to spawn new connections for each destination in the future, but it is irrelevant to this problem for now.

Note: `HAProxy` is a little different in that it still needs an ip address for the `destinationAddress` or an exception will be thrown. (`destSocketAddress` would have a null address)

https://github.com/line/armeria/blob/790a2db01a8514b591bde90932ab18f63693829e/core/src/main/java/com/linecorp/armeria/client/HAProxyHandler.java#L112

Although I think it might make more sense to validate whether the input is an ip address instead of performing a dns resolution, I've left the behavior to be unchanged for now.

Modifications:

- Now requests get `proxyConfig` and uses it to determine whether dns resolution should occur.
- `PoolKey.ipAddr` is now nullable for forward proxy requests.
- `ProxyType` has a new property `isForwardProxy`.
- Forward proxy requests now don't rely on `destinationAddress` IP for pooling connections (which was odd behavior to begin with)
- Refactored test classes s.t. inner classes in `ProxyClientIntegrationTest` are moved upwards so that other classes can test proxy behavior.

Result:

- Closes #4244

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java [hunk 1]
```diff
@@ -441,9 +442,15 @@ void invokeProxyConnectFailed(SessionProtocol protocol, PoolKey poolKey, Throwab
     }
 
     private static InetSocketAddress toRemoteAddress(PoolKey key) throws UnknownHostException {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java:444]: How about moving this method into a member method of `PoolKey` like `poolKey.toRemoteAddress()` for better encapsulation?

**Review by ikhoon — APPROVED** [review]
Thanks for the quick fix, @jrhee17! 🙇‍♂️

File: core/src/main/java/com/linecorp/armeria/client/proxy/ProxyType.java [hunk 2]
```diff
@@ -24,25 +24,38 @@ public enum ProxyType {
...
+    private final boolean isForwardProxy;
+
+    ProxyType(boolean isForwardProxy) {
+        this.isForwardProxy = isForwardProxy;
+    }
+
+    /**
+     * Returns whether this proxy is a forward proxy type.
+     */
+    public boolean isForwardProxy() {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/proxy/ProxyType.java:58]: `@UnstableApi`?

**Review by trustin — COMMENTED** [review]
Just ont nit. Nice work, @jrhee17 🙇

**Review by minwoox — APPROVED** [review]
Thanks! ❤️

**Review by trustin — APPROVED** [review]
:+1:

**Pull request merged**